### PR TITLE
History

### PR DESCRIPTION
**Pull request by andrewtookay**

### CODE REVIEW

File: app/dot/cpp/core/services/HistoryService.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    if (ValidationHelper.isEmpty(trackingId)) {
+      throw new EntityNotFoundException();
+    }
+
+    final var entity = entityService.findByField("trackingId", trackingId);
+    return getHistoryEntries(entityService, entity);
+  }
+
+  @NotNull
+  private List<HistoryEntry> getHistoryEntries(EntityService<?, ?> entityService, BaseEntity entity)
```
Comment (sircoderin) [inline app/dot/cpp/core/services/HistoryService.java:51]: Maybe rename entity cu currentState?

**Review by sircoderin — CHANGES_REQUESTED** [review]

File: app/dot/cpp/core/exceptions/FormException.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package dot.cpp.core.exceptions;
+
+import java.util.List;
+import play.data.validation.ValidationError;
+
+public class FormException extends Exception {
+
+  private final transient List<ValidationError> formErrors;
```
[the remark this answers is not in the document]
Comment (andrewtookay) [reply app/dot/cpp/core/exceptions/FormException.java:8]: because List<ValidationError> is not serializable and it was a critical code smell

File: app/dot/cpp/core/models/HistoryEntry.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  public final String comment;
+
+  public final String id;
+
+  public HistoryEntry(String userName, String dateTime, String comment, String id) {
+    this.userName = userName;
+    this.dateTime = dateTime;
+    this.comment = comment;
+    this.id = id;
+  }
```
[the remark this answers is not in the document]
Comment (andrewtookay) [reply app/dot/cpp/core/models/HistoryEntry.java:18]: It was not needed as this class is never used in a different manner, you use the constructor, set everything at creation and then you read those final variables. Its fields do not need to be set individually at any point. It is simple and clean this way.

File: app/dot/cpp/core/services/EntityService.java [hunk 4]
```diff
@@ -130,6 +148,35 @@ public int getNumberOfPages(long numEntities) {
     return (int) numberOfPages;
   }
 
+  public void save(T entity) {
+    repository.save(entity);
+  }
+
+  // sending an entityId and an entity at the same time is redundant
+  // use findByIdOrGetNewEntity where needed and replace this behaviour
```
[the remark this answers is not in the document]
Comment (andrewtookay) [reply app/dot/cpp/core/services/EntityService.java:156]: Yes! It's coming up in the next episode, the refactor PR.

File: app/dot/cpp/core/services/HistoryService.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  public static final String INVALID = "invalid";
+  private final SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm dd-MM-yyyy");
+  private final UserService userService;
+
+  @Inject
+  public HistoryService(UserService userService) {
+    this.userService = userService;
+  }
+
+  public List<HistoryEntry> getHistoryEntriesById(EntityService<?, ?> entityService, String id)
```
[the remark this answers is not in the document]
Comment (andrewtookay) [reply app/dot/cpp/core/services/HistoryService.java:30]: These <?, ?> are always extensions of BaseEntity and BaseRequest so this class can be parameterized. I can show you what I came up.

File: app/dot/cpp/core/services/HistoryService.java (hunk 1 above)
Comment (andrewtookay) [reply app/dot/cpp/core/services/HistoryService.java:51]: or currentEntity I guess?

File: app/dot/cpp/core/services/RequestErrorService.java [hunk 6]
```diff
@@ -52,6 +52,18 @@ public Result handleFormErrors(Call call, Http.Request request, Form<?> webForm)
     return getResult(call, getErrorMessage(webForm.errors(), messages));
   }
 
+  /**
+   * Handle form errors.
+   *
+   * @param call Call
+   * @param request Request
+   * @param errors List of ValidationError
+   */
```
[the remark this answers is not in the document]
Comment (andrewtookay) [reply app/dot/cpp/core/services/RequestErrorService.java:61]: I think I had the default checkstyle at the time and it was asking for this or something.

File: app/dot/cpp/core/services/HistoryService.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      throws EntityNotFoundException {
+    if (entity == null) {
+      throw new EntityNotFoundException();
+    }
+
+    final var historyEntries = new ArrayList<HistoryEntry>();
+    final var historyEntities = entityService.listHistory(entity.getTrackingId());
+
+    final var userIds =
+        historyEntities.stream().map(BaseEntity::getModifiedBy).collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (andrewtookay) [reply app/dot/cpp/core/services/HistoryService.java:61]: I renamed this to modifiedByUserIdSet.

File: app/dot/cpp/core/services/HistoryService.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+  @NotNull
+  private List<HistoryEntry> getHistoryEntries(EntityService<?, ?> entityService, BaseEntity entity)
+      throws EntityNotFoundException {
+    if (entity == null) {
+      throw new EntityNotFoundException();
+    }
+
+    final var historyEntries = new ArrayList<HistoryEntry>();
+    final var historyEntities = entityService.listHistory(entity.getTrackingId());
```
[the remark this answers is not in the document]
Comment (andrewtookay) [reply app/dot/cpp/core/services/HistoryService.java:58]: I'm not really sure about this, to me it seemed quite explicit, like those are the entities found in the history collection. I could call them oldEntities but it seems a little silly. I think it's the historyEntries and historyEntities that can cause a little confusion, but it's not like the method is that complex and hard to follow.

File: app/dot/cpp/core/services/EntityService.java [hunk 9]
```diff
@@ -200,6 +213,54 @@ public T findByIdOrGetNewEntity(String id) throws EntityNotFoundException {
...
+    final var historyStates = listHistory(currentState.getTrackingId());
+
+    final var modifiedByUserIdSet =
+        historyStates.stream().map(BaseEntity::getModifiedBy).collect(Collectors.toSet());
+    modifiedByUserIdSet.add(currentState.getModifiedBy());
+
+    final var users =
+        userRepository
+            .listWithFilter(
+                FilterBuilder.newInstance()
```
Comment (sircoderin) [inline app/dot/cpp/core/services/EntityService.java:239]: might as welll use Filters.eq here

File: app/dot/cpp/core/services/EntityService.java [hunk 10]
```diff
@@ -200,6 +213,54 @@ public T findByIdOrGetNewEntity(String id) throws EntityNotFoundException {
...
+                    .orEq(
+                        "_id",
+                        new ArrayList<>(modifiedByUserIdSet)
```
Comment (sircoderin) [inline app/dot/cpp/core/services/EntityService.java:242]: can't we loop the set without casting?

File: app/dot/cpp/core/services/EntityService.java [hunk 11]
```diff
@@ -200,6 +213,54 @@ public T findByIdOrGetNewEntity(String id) throws EntityNotFoundException {
...
+    return getHistoryEntries(entity);
+  }
+
+  @NotNull
+  private List<HistoryEntry> getHistoryEntries(BaseEntity currentState) {
+    final var historyEntries = new ArrayList<HistoryEntry>();
+    final var historyStates = listHistory(currentState.getTrackingId());
+
+    final var modifiedByUserIdSet =
+        historyStates.stream().map(BaseEntity::getModifiedBy).collect(Collectors.toSet());
```
Comment (sircoderin) [inline app/dot/cpp/core/services/EntityService.java:233]: we can use distinct instead of toSet in order to make this a list from te start

File: app/dot/cpp/core/services/EntityService.java [hunk 12]
```diff
@@ -200,6 +213,54 @@ public T findByIdOrGetNewEntity(String id) throws EntityNotFoundException {
...
+    modifiedByUserIdSet.add(currentState.getModifiedBy());
+
+    final var users =
+        userRepository
+            .listWithFilter(
+                FilterBuilder.newInstance()
+                    .orEq(
```
Comment (sircoderin) [inline app/dot/cpp/core/services/EntityService.java:240]: shouldn't this be using "in"?

File: app/dot/cpp/core/services/EntityService.java [hunk 13]
```diff
@@ -200,6 +213,54 @@ public T findByIdOrGetNewEntity(String id) throws EntityNotFoundException {
...
+                        "_id",
+                        new ArrayList<>(modifiedByUserIdSet)
+                            .stream().map(ObjectId::new).collect(Collectors.toList()))
+                    .build())
```
Comment (sircoderin) [inline app/dot/cpp/core/services/EntityService.java:244]: Might as well inline this filter
